rp2040-hal/src/timer.rs

```diff
@@ -340,6 +343,16 @@ macro_rules! impl_alarm {
                 let bits: u32 = unsafe { &*TIMER::ptr() }.armed.read().bits();
                 (bits & $armed_bit_mask) == 0
             }
+
+            /// Cancel an activated Alarm. No negative effects if it's already disabled.
+            /// Unlike `timer::cancel` trait, this only cancels the alarm and keeps the timer running
+            /// if it's already active.
+            fn cancel(&mut self) -> Result<(), ScheduleAlarmError> {
+                // safety: Only read operations are made on the timer and they should not have any UB
+                let timer = unsafe { &*TIMER::ptr() };
+
+                self.schedule_internal(timer, Instant::from_ticks(0))
```

Comment:
I think it'd be better to directly clear the relevant flag with:

```rust
unsafe {
    timer.armed.write_with_zero(|w| w.bits($armed_bit_mask));
    crate::atomic_register_access::write_bitmask_set(
        timer.intf.as_ptr(),
        $armed_bit_mask,
    );
}
```

Reply:
@ithinuel Thanks for the hint. Implemented similar to what you said but with a modification as your code was forcing the timer process the alarm even though it was not armed.

Reply:
@MuratUrsavas, it looks like your code does exactly what @ithinuel suggested. What's the modification you are mentioning? It looks like the current version still triggers the interrupt unconditionally.

Reply:
```rust
unsafe {
    timer.armed.write_with_zero(|w| w.bits($armed_bit_mask));
    crate::atomic_register_access::write_bitmask_set(       <<<<<< I'm using write_bitmask_clear instead of set.
        timer.intf.as_ptr(),
        $armed_bit_mask,
    );
}
```

As I said above, I'm using `write_bitmask_clear` instead of `write_bitmask_set`. This way it prevents forcing interrupt even it's not armed. I've tried and it wasn't triggering after arming and cancelling. (Could be any source re-triggering it?) I've proved that via tracing the SFR's step by step and then with some practical tests.

On the other hand, I've seen other unintended alarm interrupts even though the cancel method was not implemented. My guess is there could be a bug in `timer.rs` but this is just a hunch. I don't have a proof, yet.

Reply:
LOL, sorry, I swear I read the code 5 times and still missed that difference :-)

Reply:
@jannic No worries. That was like finding 7 differences in the picture :smile: